base/record_replay.h

```diff
@@ -203,6 +204,52 @@ class SCOPED_LOCKABLE AutoUnlockMaybeEventsDisallowed {
   base::Lock& lock_;
 };
 
+// This drop-in replacement of unique_ptr purposefully leaks owned memory, so as
+// to avoid unwanted cleanup operations, usually in non-deterministic execution
+// paths. First discussed here:
+// https://linear.app/replay/issue/RUN-1227/divergence-frameschedulerimpl-destroys-powermodevoter
+// Playground: https://replit.com/@Domiii/Leak-Unique-Ptr#main.cpp
+template <typename T>
+class unique_leaky_ptr {
+  std::unique_ptr<T> p;
+
+public:
+  // copy ctor
+  template <typename T_>
+  unique_leaky_ptr(T_&) = delete;
+
+  // copy assignment
+  template <typename T_>
+  unique_leaky_ptr& operator=(const T_&) = delete;
+
+  // move ctors
+  unique_leaky_ptr(unique_leaky_ptr&& u) : unique_leaky_ptr(std::move(u.p)) {}
+  template <typename T_>
+  unique_leaky_ptr(T_&& u) : p(std::move(u)) {}
+
+  // move assignments
+  unique_leaky_ptr& operator=(unique_leaky_ptr&& u) noexcept {
+    p = std::move(u.p);
+  }
+  template <typename T_>
+  unique_leaky_ptr& operator=(T_&& val) noexcept {
+    p = val;
+    return *this;
+  }
+
+  // Return the stored pointer.
+  T* operator->() const noexcept { return get(); }
+
+  /// Return the stored pointer.
+  T* get() const noexcept { return p.get(); }
+
+  // dtor
+  ~unique_leaky_ptr() {
+    // -> leak the allocated memory, before destructing `unique_ptr`.
+    p.release();
```

Comment:
This should only leak the pointer when AreEventsDisallowed, in keeping with the existing places where we leak these unique ptrs.  This keeps the non-recording case from being impacted, reduces the impact when recording/replaying, and allows us to turn on/off this leaking behavior via the record/replay features so we can see any correctness / performance impacts from behavior changes like this.  The comment on the class also needs to be updated.

Reply:
Done

Reply:
Feel free to re-review